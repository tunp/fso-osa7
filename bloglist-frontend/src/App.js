// @flow

import * as React from 'react';
import { Route } from 'react-router-dom'
import { Col, Form, FormGroup, FormControl, ControlLabel, Button } from 'react-bootstrap'
import { msgShowing } from './reducers/notificationReducer'
import { userInit } from './reducers/userReducer'
import { blogInit } from './reducers/blogReducer'
import { loginInit, login } from './reducers/loginReducer'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import Blog from './components/Blog'
import BlogList from './components/BlogList'
import BlogForm from './components/BlogForm'
import Message from './components/Message'
import Togglable from './components/Togglable'
import UsersList from './components/UsersList'
import User from './components/User'
import Menu from './components/Menu'

type Props = {
    msgShowing: (msg?: string, type?: string) => void,
    userInit: () => void,
    blogInit: () => void,
    loginInit: () => void,
    login: ({ username: string, password: string }) => void,
    user?: { name?: string }
}

type EventType = {
    preventDefault: () => void,
    target: {
        username: { value: string },
        password: { value: string }
    }
}

class App extends React.Component<Props> {
    componentDidMount = async () => {
        this.props.msgShowing() // init
        await this.props.blogInit()
        await this.props.userInit()
        this.props.loginInit()
    }

    login = async (event: EventType) => {
        event.preventDefault()
        try {
            const login_user = {
                username: event.target.username.value,
                password: event.target.password.value
            }
            await this.props.login(login_user)
            this.props.msgShowing("Logged in as " + (this.props.user && this.props.user.name ? this.props.user.name : ""), "success")
        } catch (exception) {
            this.props.msgShowing("Error: " + exception, "fail")
        }
    }

    render() {
        let body;
        if (!this.props.user) {
            body = (
                <div>
                    <h2>Kirjaudu sovellukseen</h2>
                    <Form horizontal onSubmit={this.login}>
                        <FormGroup>
                            <Col componentClass={ControlLabel} sm={2}>käyttäjätunnus</Col>
                            <Col sm={10}><FormControl name="username"/></Col>
                        </FormGroup>
                        <FormGroup>
                            <Col componentClass={ControlLabel} sm={2}>salasana</Col>
                            <Col sm={10}><FormControl type="password" name="password"/></Col>
                        </FormGroup>
                        <FormGroup>
                            <Col smOffset={2} sm={10}>
                                <Button type="submit">kirjaudu</Button>
                            </Col>
                        </FormGroup>
                    </Form>
                </div>
            )
        } else {
            body = (
                <div className={"logged_in_container"}>
                    <Menu />
                    <Route exact path="/" className={"default_route"} render={() => (
                        <div>
                            <h2>blogs</h2>
                            <BlogList />
                            <Togglable className="add_blog_togglable" button_label={"Add Blog"}>
                                <BlogForm />
                            </Togglable>
                        </div>
                    )} />
                    <Route exact path="/users" render={() => <UsersList />} />
                    <Route exact path="/users/:id" render={({match}) => <User id={match.params.id} />} />
                    <Route exact path="/blogs/:id" render={
                        ({match, history}) => <Blog
                            id={match.params.id}
                            history={history}
                        />
                    } />
                </div>
            )
        }
        return (
            <div className="container">
                <Message />
                {body}
            </div>
        );
    }
}

App.propTypes = {
    msgShowing: PropTypes.func.isRequired,
    userInit: PropTypes.func.isRequired,
    blogInit: PropTypes.func.isRequired,
    loginInit: PropTypes.func.isRequired,
    login: PropTypes.func,
    user: PropTypes.object
}

const mapStateToProps = (state) => {
    return {
        user: state.user,
    }
}

const mapDispatchToProps = {
    msgShowing,
    userInit,
    blogInit,
    loginInit,
    login
}

const ConnectedApp = connect(mapStateToProps, mapDispatchToProps)(App)

export default ConnectedApp
