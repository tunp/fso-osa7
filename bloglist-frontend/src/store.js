import { createStore, combineReducers, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import notificationReducer from './reducers/notificationReducer'
import userReducer from './reducers/userReducer'
import blogReducer from './reducers/blogReducer'
import loginReducer from './reducers/loginReducer'

const reducer = combineReducers({
    notifications: notificationReducer,
    users: userReducer,
    blogs: blogReducer,
    user: loginReducer
})

const store = createStore(reducer, applyMiddleware(thunk))

export default store
