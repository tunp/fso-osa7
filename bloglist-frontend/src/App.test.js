import React from 'react'
import { mount } from 'enzyme'
import { Provider } from 'react-redux'
import thunk from 'redux-thunk'
import renderer from 'react-test-renderer'
import { MemoryRouter, Route } from 'react-router-dom'
import store from './store'
import App from './App'
jest.mock('./services/blogs')
import blogService from './services/blogs'
jest.mock('./services/users')

import {createSerializer} from 'enzyme-to-json';
expect.addSnapshotSerializer(createSerializer({mode: 'deep'}));

describe('<App />', () => {
    let app
    describe('user not logged in', () => {
        beforeAll(() => {
            app = mount(
                <Provider store={store}>
                    <MemoryRouter initialEntries={['/']}>
                        <Route path="/" component={App} />
                    </MemoryRouter>
                </Provider>
            )
        })

        it('doesnt show any blogs', () => {
            app.update()
            expect(app.find('input[name="username"]').length).toBe(1)
            expect(app.find('input[name="password"]').length).toBe(1)
            expect(app.find('.blog_list_link').length).toBe(0)
        })
    })
    describe('user logged in', () => {
        beforeAll(async () => {
            const user = {
                _id: 101,
                username: 'testun1',
                token: '21433543',
                name: 'Test User 1'
            }
            localStorage.setItem('blogsLoggedUser', JSON.stringify(user))
            app = await mount(
                <Provider store={store}>
                    <MemoryRouter initialEntries={['/']}>
                        <Route path="/" component={App} />
                    </MemoryRouter>
                </Provider>
            )
        })

        it('shows blogs', () => {
            app.update()
            expect(app.find('input[name="username"]').length).toBe(0)
            expect(app.find('input[name="password"]').length).toBe(0)
            expect(app.find('Link.blog_list_link').length).toBe(blogService.blogs.length)
        })

        it('adds blog', async () => {
            app.update()
            const show_button = app.find('.add_blog_togglable Button.show_button')
            show_button.simulate('click')
            app.update()
            const title = app.find('input[name="title"]')
            const author = app.find('input[name="author"]')
            const url =  app.find('input[name="url"]')
            title.at(0).instance().value = "test title 3";
            title.simulate('change')
            author.at(0).instance().value = "test author 3";
            author.simulate('change')
            url.at(0).instance().value = "test url 3";
            url.simulate('change')
            // simulate click doestnt wait for the await so using addBlog directly
            //app.find('BlogForm Form').simulate('submit')
            await app.find('BlogForm').at(0).instance().addBlog({ preventDefault: ()=>{}})
            app.update()
            expect(app.find('Link.blog_list_link').length).toBe(blogService.blogs.length+1)
            // simulate click doesn't seem to work wor some reason so using history.push
            //app.find('NavLink.users_link a').simulate('click'))
            app.find('.default_route').at(0).instance().context.router.history.push('/users')
            app.update()
            var user_blogs_count = app.find('.user_blogs_count');
            let total_count = 0;
            for (let i = 0; i < user_blogs_count.length; i++) {
                total_count += user_blogs_count.at(i).instance().textContent*1
            }
            expect(total_count).toBe(blogService.blogs.length+1)
        })
    })

    describe('rendering', () => {
        let app2
        beforeAll(async () => {
            const user = {
                _id: 101,
                username: 'testun1',
                token: '21433543',
                name: 'Test User 1'
            }
            localStorage.setItem('blogsLoggedUser', JSON.stringify(user))
            app2 = await mount(
                <Provider store={store}>
                    <MemoryRouter initialEntries={['/']}>
                        <Route path="/" component={App} />
                    </MemoryRouter>
                </Provider>
            )
        })

        it('renders correctly', async () => {
            app2.update()
            expect(app2).toMatchSnapshot()
        })
    })
})
