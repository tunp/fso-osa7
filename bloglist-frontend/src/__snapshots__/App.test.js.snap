// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> rendering renders correctly 1`] = `
<div
  className="container"
>
  <div />
  <div
    className="logged_in_container"
  >
    <div
      style={
        Object {
          "backgroundColor": "#fbb",
          "padding": "5px",
        }
      }
    >
      <a
        aria-current="page"
        className="active"
        href="/"
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#f22",
            "borderRadius": "4px",
            "padding": "2px",
          }
        }
      >
        blogs
      </a>
       
      <a
        aria-current={null}
        className="users_link"
        href="/users"
        onClick={[Function]}
      >
        users
      </a>
       
      Test User 1
       logged in 
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        logout
      </button>
    </div>
    <div>
      <h2>
        blogs
      </h2>
      <ul
        className="blog_list list-group"
      >
        <li
          className="list-group-item"
        >
          <a
            className="blog_list_link"
            data-title="test title 2"
            href="/blogs/2"
            onClick={[Function]}
          >
            test title 2
             
            test author 2
          </a>
        </li>
        <li
          className="list-group-item"
        >
          <a
            className="blog_list_link"
            data-title="test title 1"
            href="/blogs/1"
            onClick={[Function]}
          >
            test title 1
             
            test author 1
          </a>
        </li>
      </ul>
      <div>
        <div
          style={
            Object {
              "display": "",
            }
          }
        >
          <button
            className="show_button btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Add Blog
          </button>
        </div>
        <div
          className="content_drop"
          style={
            Object {
              "display": "none",
            }
          }
        >
          <div>
            <form
              className="form-horizontal"
              onSubmit={[Function]}
            >
              <div
                className="form-group"
              >
                <label
                  className="col-sm-2 control-label"
                >
                  title
                </label>
                <div
                  className="col-sm-10"
                >
                  <input
                    className="form-control"
                    name="title"
                    onChange={[Function]}
                    value=""
                  />
                </div>
              </div>
              <div
                className="form-group"
              >
                <label
                  className="col-sm-2 control-label"
                >
                  author
                </label>
                <div
                  className="col-sm-10"
                >
                  <input
                    className="form-control"
                    name="author"
                    onChange={[Function]}
                    value=""
                  />
                </div>
              </div>
              <div
                className="form-group"
              >
                <label
                  className="col-sm-2 control-label"
                >
                  url
                </label>
                <div
                  className="col-sm-10"
                >
                  <input
                    className="form-control"
                    name="url"
                    onChange={[Function]}
                    value=""
                  />
                </div>
              </div>
              <div
                className="form-group"
              >
                <div
                  className="col-sm-10 col-sm-offset-2"
                >
                  <button
                    className="btn btn-default"
                    disabled={false}
                    type="submit"
                  >
                    create
                  </button>
                </div>
              </div>
            </form>
          </div>
          <button
            className="hide_button btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
